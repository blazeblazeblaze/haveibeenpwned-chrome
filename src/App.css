.App {
  margin: auto;
  height: 100vh;
  background-image: linear-gradient(to right, #ffecd2 0%, #fcb69f 100%);
  font-family: system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI",
    "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans",
    "Droid Sans", "Helvetica Neue", sans-serif;
}

.content-wrapper {
  text-align: center;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.logo-wrapper {
  display: inline-block;
  margin-bottom: 10vh;
}

.logotype {
  border-radius: 10px;
  border: 2px solid #fff!important;
  font-size: 1.4em;
  padding: 4px 8px 7px;
  margin: 7px;
  font-weight: 700;
  height: auto;
  text-decoration: none;
  color: white;
}
