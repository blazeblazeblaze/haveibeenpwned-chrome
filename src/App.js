import React, { Component } from 'react';

import api from './lib/api';
import Cache from './lib/Cache';

import Spinner from './components/Spinner';
import ErrorPage from './components/ErrorPage';
import Logo from './components/Logo';
import StatusPage from './components/StatusPage';

import spinnerSrc from './spinner.svg';
import './App.css';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoading: false,
      errors: [],
      isPwned: false,
      updatedAt: null
    };

    this.handleEmailUpdate = this.handleEmailUpdate.bind(this);
    this.handleStartOver = this.handleStartOver.bind(this);
    this.cache = new Cache('pwned_chrome');
  }

  componentDidMount() {
    const cache = this.cache.fetch();
    if (!cache) return;

    this.setState(() => {
      return cache;
    })
  }

  componentDidUpdate(prevProps, prevState) {
    const { email, isPwned } = this.state;

    if (!email) return;
    if (this.hasErrors()) return;
    if (isPwned != null) return;

    if(!this.isExpired()) {
      this.setState(() => {
        return {...this.state,
          isPwned: false,
          errors: ['API rate limit'],
          isLoading: false,
          email: null
        }
      })
    } else {
      api.fetchBreaches(email)
        .then((breaches) => {
          this.setState(() => {
            return {...this.state,
              isPwned: breaches.length > 0,
              errors: [],
              isLoading: false,
              updatedAt: new Date()
            };
          });
        })
        .then(() => {
          return this.cache.store(this.state);
        })
        .catch((error) => {
          this.setState(() => {
            return {...this.state,
              isPwned: false,
              errors: ['server error'],
              isLoading: false,
              updatedAt: new Date()
            };
          });
        });
    }
  }

  handleStartOver(e) {
    e.preventDefault();

    this.setState(() => {
      return {...this.state,
        isLoading: false,
        errors: [],
        isPwned: null,
        email: null
      }
    });
  }

  handleEmailUpdate(email) {
    this.setState(() => {
      return {...this.state,
        email: email,
        isLoading: true,
        isPwned: null
      }
    });
  }

  // Where does this go?
  isExpired() {
    const { updatedAt } = this.state;
    return new Date() - new Date(updatedAt) > 3000; // 3s
  }

  hasErrors() {
    const { errors } = this.state;
    return !this.isExpired() && errors && errors.length > 0;
  }
  // Where does this go?

  render() {
    const { isLoading, email, isPwned, errors } = this.state;

    return(
      <div className='App pure-g'>
        <div className="pure-u-1-1 content-wrapper">
          <Logo onStartOver={this.handleStartOver} />

          {this.hasErrors() &&
            <ErrorPage errors={errors} onStartOver={this.handleStartOver} />}

          {!this.hasErrors() && !isLoading && <StatusPage
            isPwned={isPwned}
            email={email}
            onStartOver={this.handleStartOver}
            onSubmit={this.handleEmailUpdate} />}

          {isLoading && <Spinner src={spinnerSrc} />}
        </div>
      </div>
    )
  }
}

export default App;
