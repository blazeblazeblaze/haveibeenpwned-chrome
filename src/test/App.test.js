import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import renderer from 'react-test-renderer';

beforeAll(() => {
  const localStorageMock = {
    getItem: jest.fn(),
    setItem: jest.fn(),
    clear: jest.fn()
  };

  global.localStorage = localStorageMock;
});

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

jest.mock('../components/Spinner', () => () => 'SpinnerComponent');
jest.mock('../components/ErrorPage', () => () => 'ErrorPageComponent');
jest.mock('../components/Logo', () => () => 'LogoComponent');
jest.mock('../components/StatusPage', () => () =>  'StatusPageComponent');

test('App renders - basic', () => {
  const component = renderer.create(
    <App />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
