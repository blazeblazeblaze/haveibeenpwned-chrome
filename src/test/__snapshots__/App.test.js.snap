// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders - basic 1`] = `
<div
  className="App pure-g"
>
  <div
    className="pure-u-1-1 content-wrapper"
  >
    LogoComponent
    StatusPageComponent
  </div>
</div>
`;
